### ISSUE TITLE
[ENH] Explore using a Cython-specific variant of PEP-484 "type comments" to specify types and function signatures?

### ISSUE THREAD

**Issue by ermo** [cython/cython#4030]
**Alternative specification syntax for simple C types in Cython**

Being able to make use of the Python type-annotation support in Cython pure mode is very convenient.  

Currently, the [cython-3.0a6 documentation for Cython pure mode](https://cython.readthedocs.io/en/latest/src/tutorial/pure.html#pep-484-type-annotations) shows the following example for
simple types:

```cython
import cython

def func():
    # Cython types are evaluated as for cdef declarations
    x: cython.int               # cdef int x
    y: cython.double = 0.57721  # cdef double y = 0.57721
    z: cython.float = 0.57721   # cdef float z  = 0.57721

    # Python types shadow Cython types for compatibility reasons
    a: float = 0.54321          # cdef double a = 0.54321
    b: int = 5                  # cdef object b = 5
    c: long = 6                 # cdef object c = 6
    pass
    
```

What if the the Cython pure mode could be extended to support Cython-specific [PEP-484 type comments](https://www.python.org/dev/peps/pep-0484/#type-comments) by super-imposing a more specific Cython type on top of the CPython type annotation like so:

```cython
import cython

def func():
    # Cython types are evaluated as for cdef declarations if a
    # `# ctype: ` type comment is used
    x: int                      # ctype: cdef int x
    y: float = 0.57721          # ctype: cdef double y
    z: float = 0.57721          # ctype: cdef float z

    # Python types shadow Cython types for compatibility reasons
    # (note no `# ctype: ` type comment here)
    a: float = 0.54321          # cdef double a = 0.54321
    b: int = 5                  # cdef object b = 5
    c: long = 6                 # cdef object c = 6
    pass
```

**Alternative specification for .pxd function signatures in Cython**

This is the original non-typed Python example from the same Cython documentation page in the [Augmenting .pxd section](https://cython.readthedocs.io/en/latest/src/tutorial/pure.html#augmenting-pxd):

```cython
def myfunction(x, y=2):
    a = x - y
    return a + x * y

def _helper(a):
    return a + 1

class A:
    def __init__(self, b=0):
        self.a = 3
        self.b = b

    def foo(self, x):
        print(x + _helper(1.0))
```

Function signature definitions are usually added in .pxd files or as (somewhat cumbersome) decorators. But what if perhaps the `# ctype: ` type-comments could also be used for super-imposing more specific Cython function signatures when added to a Python type annotated version of the above:

```cython
import cython

def myfunction(x: int, y: int = 2) -> int:  # ctype: cpdef int myfunction(int x, int y=*)
    a = x - y
    return a + x * y

def _helper(a: int) -> int:  # ctype: cdef double _helper (double a)
    return a + 1

class A: # ctype: cdef Class A
    def __init__(self, b: int = 0):  # ctype: cdef public int a, b
        self.a = 3
        self.b = b

    def foo(self, x: float):  # ctype: cpdef foo(self, double x)
        print(x + _helper(1.0))
```

**Alternatives considered**

None. 

The goal was to build on the existing PEP-484 type comments specification in a clear and easy way while only making a single concise change from `# type: ` to `# ctype: `.

The rationale is that since the Cython-specific type definitions are hidden behind a comment, such definitions are not expected to conflict with code executed in the CPython interpreter, type annotated or not.

At the same time, the `c`-prefix in `# ctype: ` is easily understood as a native C type used by Cython and is thus not expected to conflict with existing `# type:` parsers.

Note also that the proposal re-uses the existing Cython cdef/cpdef syntax in the type comments.

Finally, note that without the Cython-specific `# ctype: ` type comments, Cython would simply fall back to the existing pure behaviour, which suggests that this proposal could represent a backwards-compatible change.

**Context**

The only goal with this proposal is to get a feel for whether the suggestion above is feasible.

**Comment by scoder** [comment]
Repeating the type violates the DRY principle.
Could be that you also misinterpreted the example in the docs. The comments are … comments. They don't have a special meaning.

**Comment by ermo** [comment]
@scoder 

So, the basic premise here was to be able to use normal python type annotations and then being able to super-impose cython-specif type information on top.

How would one do that with cython as it stands?

**Comment by scoder** [comment]
Why not just use Cython types? Python type analysers can understand them.

**Comment by da-woods** [comment]
My view would be:
* Cython probably already has "too many" ways to specify types - they're all there for good reasons but there's three or four different ways and it can be hard to follow what overrides what at times. I'd be keen to avoid adding another.
* I'm not convinced that "Python `int` does not got to C `int`" is a real problem (as Scoder says, `cython.int` works fine here), but if I had to solve it then the right way to do it would probably be to add a compiler directive (and otherwise leave the typing code untouched)

**Comment by ermo** [comment]
Let me rephrase the question then:

If I specify any of the `cython.long/cython.int/etc.` types, will they all map to CPython `int` type-annotation in a many-to-one way when run under CPython-compatible type checkers?

**Comment by da-woods** [comment]
> If I specify any of the `cython.long/cython.int/etc.` types, will they all map to CPython `int` type-annotation in a many-to-one way when run under CPython-compatible type checkers?

They should do - Shadow.py defines them like that for when you run `import cython` in regular Python

```
>>> import cython
>>> cython.int
int
```

(by design, in Shadow.py).

**Comment by ermo** [comment]
First of all, thanks for engaging with me.  I realise that I should probably have to spent a little more time describing the motivation for the examples I gave, but I can't change that now.

My starting point was that I wondered how to make cython pure mode as useful and non-intrusive for me (and others) when invoking the cython compiler on CPython code with only bog standard CPython type hints.

I (obviously) consider it a huge engineering boon to be able to conveniently re-use the existing CPython type hints to speed up modules with cython pure mode without having to spend too much time re-specifying types unless deemed absolutely necessary for extra optimisation of hot code paths.

I also obviously respect the judgement that adding another type-annotation mechanism is not feasible/desirable.

Having thought about it some more, it indeed seems to me that having a cython flag that -- when specified -- auto-translates CPython `int` and `float` type hints to the native machine type for each (e.g. `cython.long` and `cython.double` on AMD64) is probably the easiest/simplest way forward in terms of easy speedups when using cython pure mode on code with purely CPython type hints?

That way, if a user specifies said flag, they could be warned about the implications and be told that if something breaks due to differing semantics between cython and CPython `int` and `float` types, they get to keep the pieces?

Then, in the cases where it makes sense to switch to more specific types (`int` -> `cython.int` or `float` -> `cython.float` for instance), the user can do that incrementally as benchmarks indicate?
